### PR TITLE
Updating Katas to 0.6

### PR DESCRIPTION
**Pull request by bettinaheim**
This PR updates all katas to the 0.6 release. 
The package references will be set to a suitable version after the release.

### CODE REVIEW

File: DeutschJozsaAlgorithm/Tests.qs [hunk 1]
```diff
@@ -111,9 +107,9 @@ namespace Quantum.Kata.DeutschJozsaAlgorithm {
         
         // the mask with only the K-th element set to 1 corresponds to Oracle_Kth_Qubit
         for (i in 0 .. L - 1) {
-            set r[i] = 1;
+            set r w/= i <- 1;
```
Comment (cgranade) [inline DeutschJozsaAlgorithm/Tests.qs]: I think this `set` statement and the one at L112 can both be removed entirely in favor of replacing `r` in the next line by `r w/ i <- 1`.

File: GroversAlgorithm/Tests.qs [hunk 2]
```diff
@@ -59,7 +56,7 @@ namespace Quantum.Kata.GroversAlgorithm {
             let pattern = BoolArrFromPositiveInt(RandomIntPow2(n), n);
             let testOp = QubitArrayWrapperOperation(Oracle_ArbitraryPattern(_, _, pattern), _);
             let refOp = QubitArrayWrapperOperation(Oracle_ArbitraryPattern_Reference(_, _, pattern), _);
-            AssertOperationsEqualReferenced(testOp, refOp, n + 1);
+            AssertOperationsEqualReferenced(n+1, testOp, refOp);
```
Comment (cgranade) [inline GroversAlgorithm/Tests.qs]:
```suggestion
            AssertOperationsEqualReferenced(n + 1, testOp, refOp);
```

File: JointMeasurements/Tests.qs [hunk 3]
```diff
@@ -50,29 +50,26 @@ namespace Quantum.Kata.JointMeasurements {
...
-        body (...) {
-            // prep cos(alpha) * |0..0⟩ + sin(alpha) * |1..1⟩
-            Ry(2.0 * alpha, qs[0]);
-            for (i in 1 .. Length(qs) - 1) {
-                CNOT(qs[0], qs[i]);
-            }
+        // prep cos(alpha) * |0..0⟩ + sin(alpha) * |1..1⟩
+        Ry(2.0 * alpha, qs[0]);
+        for (i in 1 .. Length(qs) - 1) {
+            CNOT(qs[0], qs[i]);
```
Comment (cgranade) [inline JointMeasurements/Tests.qs]: Minor note, but this can be written as `ApplyToEachA(CNOT(Head(qs), _), Rest(qs));`.

File: Measurements/ReferenceImplementation.qs [hunk 4]
```diff
@@ -429,36 +431,23 @@ namespace Quantum.Kata.Measurements {
...
             let result = M(q);
-            if (result == One) {
-                // this can only arise if the state was |+⟩
-                set output = 1;
-            }
-            else {
-                set output = -1;
-            }
+            // result is One only if the state was |+⟩
+            return result == One ? 1 | -1;
```
Comment (cgranade) [inline Measurements/ReferenceImplementation.qs]: 💕

File: SolveSATWithGrover/ReferenceImplementation.qs [hunk 5]
```diff
@@ -10,171 +10,136 @@
...
-                // To calculate the final answer, apply the And oracle with the ancilla qubits as inputs
-                Oracle_And_Reference([a1, a2], target);
-                // Uncompute the values of the ancillas before releasing them (no measuring!)
-                Adjoint WithA(ApplyToEachA(X, _), Oracle_Or_Reference(_, a2), queryRegister);
-                Adjoint Oracle_Or_Reference(queryRegister, a1);
-            }
+    operation Oracle_Xor_2SAT (queryRegister : Qubit[], target : Qubit) : Unit
+    is Adj {        
+        // x₀ ⊕ x₁ = (x₀ ∨ x₁) ∧ (¬x₀ ∨ ¬x₁)
+        // Allocate 2 ancilla qubits to store results of clause evaluation
```
Comment (cgranade) [inline SolveSATWithGrover/ReferenceImplementation.qs]:
```suggestion
        // Allocate 2 auxillary qubits to store results of clause evaluation
```

File: SolveSATWithGrover/Tests.qs [hunk 6]
```diff
@@ -126,7 +123,7 @@ namespace Quantum.Kata.GroversAlgorithm {
         for (n in 2 .. 5) {
             AssertOracleImplementsFunction(n, Oracle_AlternatingBits, AlternatingBits);
 
-            AssertOperationsEqualReferenced(testOp, refOp, n + 1);
+            AssertOperationsEqualReferenced(n+1, testOp, refOp);
```
Comment (cgranade) [inline SolveSATWithGrover/Tests.qs]:
```suggestion
            AssertOperationsEqualReferenced(n + 1, testOp, refOp);
```

File: DeutschJozsaAlgorithm/ReferenceImplementation.qs [hunk 7]
```diff
@@ -356,19 +319,22 @@ namespace Quantum.Kata.DeutschJozsaAlgorithm {
             
             // now y = Σᵢ 𝑟ᵢ
             
+            // Declare a Bool array in which the result will be stored;
```
Comment (tcNickolas) [inline DeutschJozsaAlgorithm/ReferenceImplementation.qs]:
```suggestion
            // Declare an Int array in which the result will be stored;
```

File: GHZGame/ReferenceImplementation.qs [hunk 8]
```diff
@@ -84,19 +82,18 @@ namespace Quantum.Kata.GHZGame {
...
-        mutable abc = new Bool[3];
 
         using (qs = Qubit[3]) {
             CreateEntangledTriple_Reference(qs);
-
+            
+            mutable abc = new Bool[3];
             for (i in 0..2) {
-                set abc[i] = strategies[i](qs[i]);
+                set abc w/= i <- strategies[i](qs[i]);
```
Comment (tcNickolas) [inline GHZGame/ReferenceImplementation.qs]: @cgranade I think this was on of the places where I was looking for a library function to do it: apply i-th element of one array to i-th element of another and put the result in i-th element of the answer.

File: Measurements/ReferenceImplementation.qs [hunk 9]
```diff
@@ -468,7 +457,7 @@ namespace Quantum.Kata.Measurements {
     //        |A⟩ = 1/sqrt(2) (|0⟩ + |1⟩),
     //        |B⟩ = 1/sqrt(2) (|0⟩ + ω |1⟩),
     //        |C⟩ = 1/sqrt(2) (|0⟩ + ω² |1⟩).
-    //	      where ω = exp(2π/3) denotes a primitive, complex 3rd root of unity.
+    //          where ω = exp(2π/3) denotes a primitive, complex 3rd root of unity.
```
Comment (tcNickolas) [inline Measurements/ReferenceImplementation.qs]: I wonder what is the reason for this change?

File: JointMeasurements/Tests.qs (hunk 3 above)
Comment (bettinaheim) [reply JointMeasurements/Tests.qs]: There are plenty of things like that that could make use of the standard libraries that I did not address with this PR. This one was meant to do the bare minimum of updates related to breaking changes since we needed it for the release. I ended up also doing a little bit of refactoring to take advantage of new language features (couldn't help it), but I didn't do any refactoring related to leveraging the standard libraries. Let's make that a separate PR.

File: SimonsAlgorithm/Tasks.qs [hunk 10]
```diff
@@ -86,18 +76,15 @@ namespace Quantum.Kata.SimonsAlgorithm {
...
+    operation Oracle_MultidimensionalOperatorOutput (x : Qubit[], y : Qubit[], A : Int[][]) : Unit
+    is Adj {
         
-        body (...) {
-            // The following lines enforce the constraints on the input arrays.
-            // You don't need to modify it. Feel free to remove it, this won't cause your code to fail.
-            AssertIntEqual(Length(x), Length(A[0]), "Arrays x and A[0] should have the same length");
-            AssertIntEqual(Length(y), Length(A), "Arrays y and A should have the same length");
+        // The following lines enforce the constraints on the input arrays.
+        // You don't need to modify it. Feel free to remove it, this won't cause your code to fail.
```
Comment (tcNickolas) [inline SimonsAlgorithm/Tasks.qs]:
```suggestion
        // You don't need to modify them. Feel free to remove them, this won't cause your code to fail.
```

File: SimonsAlgorithm/Tests.cs [hunk 11]
```diff
@@ -46,20 +46,16 @@ public Instance(List<List<long>> transformation, List<long> kernel, int instance
...
             {
                 get
                 {
-                    var array = new QArray<QArray<long>>(
-                        transformation.Select(
-                            vector => new QArray<long>(vector)))
-                    {
-                        new QArray<long>(transformation.Last())
-                    };
+                    var array = (IQArray<IQArray<long>>)new QArray<IQArray<long>>(transformation.Select(vector => new QArray<long>(vector)));
```
Comment (tcNickolas) [inline SimonsAlgorithm/Tests.cs]: Could you please keep this split into multiple lines for readabilty?

File: DeutschJozsaAlgorithm/Tests.qs (hunk 1 above)
Comment (bettinaheim) [reply DeutschJozsaAlgorithm/Tests.qs]: You are right of course - missed that. Done.

File: CHSHGame/Tests.qs [hunk 12]
```diff
@@ -38,16 +38,16 @@ namespace Quantum.Kata.CHSHGame {
                 set wins = wins + 1;
             }
         }
-        Message($"Win rate {ToDouble(wins) / 1000.}");
+        Message($"Win rate {IntAsDouble(wins) / 1000.}");
 
-        AssertBoolEqual(wins >= 700, true,
+        EqualityFactB(wins >= 700, true,
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply CHSHGame/Tests.qs]: If you don't mind, there are currently plenty of EqualityFactB (because that was the renaming suggestions for hte existing code) everywhere, so let's address that as part of a separate pass over all katas to incorporate features of the standard libraries when appropriate.

File: BasicGates/ReferenceImplementation.qs [hunk 13]
```diff
@@ -72,27 +72,19 @@ namespace Quantum.Kata.BasicGates {
...
     //        If the qubit is in superposition, change its state according to the effect on basis vectors.
-    operation AmplitudeChange_Reference (q : Qubit, alpha : Double) : Unit {
-        
-        body (...) {
-            Ry(2.0 * alpha, q);
-        }
-        
-        adjoint invert;
+    operation AmplitudeChange_Reference (q : Qubit, alpha : Double) : Unit
+    is Adj {
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply BasicGates/ReferenceImplementation.qs]: I can do that, just keep in mind that 
operation Foo () : ((Unit => Unit) is Adj) {
...
}
is not the same as 
operation Foo () : (Unit => Unit) is Adj {
...
}
To avoid that confusion, the current formatting is to put it on a new line. We can have a chat about that offline.

File: DeutschJozsaAlgorithm/ReferenceImplementation.qs [hunk 14]
```diff
@@ -23,15 +24,12 @@ namespace Quantum.Kata.DeutschJozsaAlgorithm {
...
-        body (...) {
-            // Since f(x) = 0 for all values of x, |y ⊕ f(x)⟩ = |y⟩.
-            // This means that the operation doesn't need to do any transformation to the inputs.
-            // Build the project and run the tests to see that T01_Oracle_Zero_Test test passes.
-        }
-        
-        adjoint invert;
+    operation Oracle_Zero_Reference (x : Qubit[], y : Qubit) : Unit
+    is Adj {
+
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply DeutschJozsaAlgorithm/ReferenceImplementation.qs]: done

File: BasicGates/BasicGates.ipynb [hunk 15]
```diff
@@ -202,13 +202,9 @@
...
     "%kata T14_AmplitudeChange_Test\n",
     "\n",
-    "operation AmplitudeChange (q : Qubit, α : Double) : Unit {\n",
+    "operation AmplitudeChange (q : Qubit, α : Double) : Unit is Adj {\n",
     "\n",
-    "    body (...) {\n",
-    "        // ...\n",
-    "    }\n",
-    "\n",
-    "    adjoint invert;\n",
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply BasicGates/BasicGates.ipynb]: Thanks, I missed that!

File: azure-pipelines.yml [hunk 16]
```diff
@@ -15,4 +15,4 @@ pool:
   vmImage: 'VS2017-Win2016'
 
 steps:
-- template: scripts/step-build-katas.yml
\ No newline at end of file
+- template: scripts/step-build-katas.yml
```
Comment (tcNickolas) [inline azure-pipelines.yml]: Please exclude this file from the changes if it is whitespace-only change. Did you mean to add a step validate-notebooks to it instead?

File: GHZGame/ReferenceImplementation.qs (hunk 8 above)
Comment (bettinaheim) [reply GHZGame/ReferenceImplementation.qs]: @cgranade If such a library function/operation does not exist, that would be a good one to add the the M.Q.Array.

File: utilities/Microsoft.Quantum.Katas/CheckKataMagic.cs [hunk 17]
```diff
@@ -0,0 +1,203 @@
...
+            this.Kind = SymbolKind.Magic;
+            this.Execute = this.Run;
+
+            this.Resolver = resolver;
+            this.Compiler = compiler;
+            this.Logger = logger;
+        }
+
+        /// <summary>
+        /// The Resolver let's us find compiled Q# operations from the workspace
```
Comment (tcNickolas) [inline utilities/Microsoft.Quantum.Katas/CheckKataMagic.cs]:
```suggestion
        /// The Resolver lets us find compiled Q# operations from the workspace
```

File: Measurements/ReferenceImplementation.qs (hunk 9 above)
Comment (bettinaheim) [reply Measurements/ReferenceImplementation.qs]: Ah, I just did a regex replacement of all tabs in favor of 4 spaces.

File: utilities/Microsoft.Quantum.Katas/CheckKataMagic.cs [hunk 18]
```diff
@@ -0,0 +1,203 @@
...
+        /// <summary>
+        /// The list of user-defined Q# code snippets from the notebook.
+        /// </summary>
+        protected ICompilerService Compiler { get; }
+
+        protected ILogger<KataMagic> Logger { get; }
+
+        /// <summary>
+        /// What this Magic does when triggered. It will:
+        /// - find the Kata to execute based on the Kata name,
```
Comment (tcNickolas) [inline utilities/Microsoft.Quantum.Katas/CheckKataMagic.cs]: Does this work with katas (whole tutorials) or with tasks (individual tasks within a tutorial)? I think it's better to disambiguate right away

File: Measurements/ReferenceImplementation.qs (hunk 9 above)
Comment (bettinaheim) [reply Measurements/ReferenceImplementation.qs]: I made it align again with the three lines above.

File: utilities/Microsoft.Quantum.Katas/CheckKataMagic.cs [hunk 19]
```diff
@@ -0,0 +1,203 @@
...
+            catch (Exception e)
+            {
+                Logger?.LogWarning(e, "Unexpected error.");
+                channel.Stderr(e.Message);
+                return null;
+            }
+        }
+
+        /// <summary>
+        /// Executes the given kata using the provided <c>userAnswer</c> as the actual answer.
```
Comment (tcNickolas) [inline utilities/Microsoft.Quantum.Katas/CheckKataMagic.cs]: Same question - I suspect this refers to a single task, not a whole kata

File: DeutschJozsaAlgorithm/DeutschJozsaAlgorithm.sln [hunk 20]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
-# Visual Studio 15
-VisualStudioVersion = 15.0.27130.2036
+# Visual Studio Version 16
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply DeutschJozsaAlgorithm/DeutschJozsaAlgorithm.sln]: sure, done.

File: SimonsAlgorithm/Tests.cs (hunk 11 above)
Comment (bettinaheim) [reply SimonsAlgorithm/Tests.cs]: sure

File: azure-pipelines.yml (hunk 16 above)
Comment (bettinaheim) [reply azure-pipelines.yml]: Remove the whitespace change. 
While Andres has put together a script to verify notebooks, I don't think we are quite ready yet to add a suitable build step.

File: utilities/Microsoft.Quantum.Katas/CheckKataMagic.cs (hunk 18 above)
Comment (bettinaheim) [reply utilities/Microsoft.Quantum.Katas/CheckKataMagic.cs]: @anpaz-msft Could you edit that as appropriate?

File: utilities/Microsoft.Quantum.Katas/CheckKataMagic.cs (hunk 19 above)
Comment (bettinaheim) [reply utilities/Microsoft.Quantum.Katas/CheckKataMagic.cs]: @anpaz-msft  same here - thanks!

**Comment by anpaz-msft** [comment]
LGTM now. Waiting on QDK's CI build to confirm everything is ok.

File: BasicGates/ReferenceImplementation.qs (hunk 13 above)
[the remark this answers is not in the document]
Comment (tcNickolas) [reply BasicGates/ReferenceImplementation.qs]: I'll be happy to chat offline on Monday. I think that in the katas the chances for this kind of confusion are extremely rare, and in the `Tasks.qs` files, especially in the BasicGates kata where we don't have any operations returning operations, having `is Adj` instead of `body` has a higher chance to confuse the reader.
